File: python week 2/assignment6.py
# Simple Python 3 program to add two
# polynomials

# A utility function to return maximum 
# of two integers

# A[] represents coefficients of first polynomial
# B[] represents coefficients of second polynomial
# m and n are sizes of A[] and B[] respectively
def add(A, B, m, n):

    size = max(m, n);
    sum = [0 for i in range(size)]

    # Initialize the product polynomial
    
    for i in range(0, m, 1):
        sum[i] = A[i]

    # Take ever term of first polynomial
    for i in range(n):
        sum[i] += B[i]

    return sum

# A utility function to print a polynomial
def printPoly(poly, n):
    for i in range(n):
        print(poly[i], end = "")
        if (i != 0):
            print("x^", i, end = "")
        if (i != n - 1):
            print(" + ", end = "")

# Driver Code
if __name__ == '__main__':
    
    # The following array represents
    # polynomial 5 + 10x^2 + 6x^3
    A = [5, 0, 10, 6]

    # The following array represents
    # polynomial 1 + 2x + 4x^2
    B = [1, 2, 4]
    m = len(A)
    n = len(B)

    print("First polynomial is")
    printPoly(A, m)
    print("\n", end = "")
    print("Second polynomial is")
    printPoly(B, n)
    print("\n", end = "")
    sum = add(A, B, m, n)
    size = max(m, n)

    print("sum polynomial is")
    printPoly(sum, size)
    
# This code is contributed by
# Sahil_Shelangia

File: c week 2/assignment6.c
#include <stdio.h>

#define MAX 100  // Maximum degree + 1

// Function to add two polynomials
void addPolynomials(int A[], int B[], int m, int n, int sum[]) {
    int size = (m > n) ? m : n;

    // Initialize the result with 0s
    for (int i = 0; i < size; i++) {
        sum[i] = 0;
    }

    // Copy A[] to sum[]
    for (int i = 0; i < m; i++) {
        sum[i] = A[i];
    }

    // Add B[] to sum[]
    for (int i = 0; i < n; i++) {
        sum[i] += B[i];
    }
}

// Function to print a polynomial
void printPolynomial(int poly[], int n) {
    for (int i = 0; i < n; i++) {
        printf("%d", poly[i]);
        if (i != 0) {
            printf("x^%d", i);
        }
        if (i != n - 1) {
            printf(" + ");
        }
    }
    printf("\n");
}

int main() {
    // Polynomial A: 5 + 0x + 10x^2 + 6x^3
    int A[] = {5, 0, 10, 6};
    // Polynomial B: 1 + 2x + 4x^2
    int B[] = {1, 2, 4};

    int m = sizeof(A) / sizeof(A[0]);
    int n = sizeof(B) / sizeof(B[0]);

    int size = (m > n) ? m : n;
    int sum[MAX];

    printf("First polynomial is:\n");
    printPolynomial(A, m);

    printf("Second polynomial is:\n");
    printPolynomial(B, n);

    addPolynomials(A, B, m, n, sum);

    printf("Sum polynomial is:\n");
    printPolynomial(sum, size);

    return 0;
}
